contracts/UniswapV3Staker.sol

```diff
@@ -291,25 +297,33 @@ contract UniswapV3Staker is
 
         // TODO: is SafeMath necessary here? Could we do just a subtraction?
         incentives[incentiveId].totalRewardUnclaimed = uint128(
-            SafeMath.sub(incentives[incentiveId].totalRewardUnclaimed, reward)
+            SafeMath.sub(incentive.totalRewardUnclaimed, reward)
         );
 
-        /* TODO: This will be fixed in https://github.com/omarish/uniswap-v3-staker/issues/38
-        so that collecting and unstaking are two separate functions */
-        try
-            IERC20Minimal(incentives[incentiveId].rewardToken).transfer(
-                params.to,
-                reward
-            )
-        returns (bool) {
-            // It didn't fail
-        } catch {
-            // It failed
-        }
+        rewards[incentive.rewardToken][msg.sender] = uint128(
+            SafeMath.add(rewards[incentive.rewardToken][msg.sender], reward)
+        );
 
         emit TokenUnstaked(params.tokenId);
     }
 
+    /// @inheritdoc IUniswapV3Staker
+    function claimReward(address rewardToken, address to) external override {
+        uint128 reward = rewards[rewardToken][msg.sender];
+
+        require(
+            IERC20Minimal(rewardToken).transfer(
+                to,
+                reward
+            ),
+            'REWARD_TRANSFER_FAILED'
+        );
+
+        rewards[rewardToken][msg.sender] = 0;
```

Comment:
put this state change before `transfer()`, or else a malicious token could do a sort of reentrancy attack and drain its funds.